### PR TITLE
chore(codecs): Implement `character_delimited` framing

### PR DESCRIPTION
**Pull request by pablosichert**
Part of #7463, has been split off from #8236.

### CODE REVIEW

File: src/sources/util/tcp_error.rs [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+use tokio_util::codec::LinesCodecError;
+
+/// An error that occurs in the context of TCP connections.
+pub trait TcpError {
+    /// The TCP error has been fatal, the TCP stream should no longer be read
+    /// since it will hang up indefinitely.
+    ///
+    /// This can occur e.g. when reading the header of a length-delimited codec
+    /// failed and it can no longer be determined where the next header starts.
+    fn is_fatal(&self) -> bool;
```
Comment (bruceg) [inline src/sources/util/tcp_error.rs:12]: Would it be worth making these semantics more explicit in the function naming? I initially mistook `is_fatal` for meaning the operation could be retried. Maybe the opposite would work — `can_continue`?

File: src/codec/framers/character_delimited.rs [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+use std::{cmp, io, usize};
+use tokio_util::codec::{Decoder, Encoder};
+
+/// Config used to build a `CharacterDelimitedCodec`.
+#[derive(Debug, Clone, Default, Deserialize, Serialize)]
+pub struct CharacterDelimitedDecoderConfig {
+    /// The character that delimits byte sequences.
+    delimiter: char,
+    /// The maximum length of the byte buffer.
+    max_length: Option<usize>,
```
Comment (bruceg) [inline src/codec/framers/character_delimited.rs:13]: It should be made explicit that (as far as I can tell from the code below) this length does *not* include the trailing delimiter.

File: src/codec/framers/character_delimited.rs [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+        let frame = match self.decode(buf)? {
+            Some(frame) => Some(frame),
+            None if !buf.is_empty() && !self.is_discarding => {
+                let frame = buf.split_to(buf.len());
+                self.next_index = 0;
+
+                Some(frame.into())
+            }
+            _ => None,
+        };
```
Comment (bruceg) [inline src/codec/framers/character_delimited.rs:150]: Can we use adapters here, or does it cause borrow check issues?

```suggestion
        let frame = self.decode(buf)?
            .or_else(|| {
                (!buf.is_empty() && !self.is_discarding).then(|| {
                    self.next_index = 0;
                    buf.split_to(buf.len()).into()
                }
            });
```

File: src/sources/util/tcp_error.rs (hunk 1 above)
Comment (pablosichert) [reply src/sources/util/tcp_error.rs]: Adressed in 404e275986e7e05c54205bd8beb651239904d167

File: src/codec/framers/character_delimited.rs (hunk 3 above)
Comment (pablosichert) [reply src/codec/framers/character_delimited.rs]: Changed in 40183c2915280331443ff08fb25e968911d20748

File: src/codec/framers/character_delimited.rs (hunk 2 above)
Comment (pablosichert) [reply src/codec/framers/character_delimited.rs]: Added in 935256e0787542222b664cdba1cb4aa4b3f31d7f

**Comment by pablosichert** [comment]
There have been CI failures, but they're unrelated to these changes (in `sources::socket::test::tcp_with_tls_intermediate_ca`).

File: src/codecs/framers/character_delimited.rs [hunk 4]
```diff
@@ -0,0 +1,315 @@
...
+
+#[typetag::serde(name = "character_delimited")]
+impl FramingConfig for CharacterDelimitedDecoderConfig {
+    fn build(&self) -> crate::Result<BoxedFramer> {
+        Ok(Box::new(
+            self.max_length
+                .map(|max_length| {
+                    CharacterDelimitedCodec::new_with_max_length(self.delimiter, max_length)
+                })
+                .unwrap_or_else(|| CharacterDelimitedCodec::new(self.delimiter)),
```
Comment (bruceg) [inline src/codecs/framers/character_delimited.rs:26]: Personally, I'd go for a `match`, but that's hardly worth blocking it for.
```suggestion
            match self.max_length {
                Some(max_length) => CharacterDelimitedCodec::new_with_max_length(self.delimiter, max_length),
                None => CharacterDelimitedCodec::new(self.delimiter),
            }
```

**Review by bruceg — APPROVED** [review]
Just one trivial non-blocking suggestion, otherwise LGTM

File: src/codecs/framers/character_delimited.rs (hunk 4 above)
Comment (pablosichert) [reply src/codecs/framers/character_delimited.rs:26]: I think I changed it to `map` based on your suggestion in https://github.com/timberio/vector/pull/8236#discussion_r688239076 😄

Personally, I also prefer `match` - but do you have a heuristic when you favor the one over the other? Looking at the example, I assume when passing in the function without an explicit closure is possible?

Comment (pablosichert) [reply src/codecs/framers/character_delimited.rs]: Changed in b220fada1f5a51f2f01c52027271a4db84593b30

Comment (bruceg) [reply src/codecs/framers/character_delimited.rs]: Well, I think it's fair to say that reviewer was a fool. This is *obviously* the one true way of doing things. 😛 😢 In my defense, that was a little different pattern, where it looked like the `map` part could be reduced down to a function reference.

My heuristic for preferring one over the other is pretty much that of simplicity — whichever form results in the simplest expression is probably the best, partially guided by `rustfmt`, with a thumb on the scale towards the more functional form.

**Pull request merged**